packages/slate/src/interfaces/editor.ts

```diff
@@ -1049,12 +1058,13 @@ export const Editor = {
 
       if (Text.isText(node)) {
         const isFirst = Path.equals(path, first.path)
-        available = node.text.length
-        offset = reverse ? available : 0
 
         if (isFirst) {
-          available = reverse ? first.offset : available - first.offset
-          offset = first.offset
+          available = reverse ? first.offset : node.text.length - first.offset
```

Comment:
These 5 lines don't appear to be logically different from what came before. Why the change?

Reply:
No they aren't.  I just thought it expressed the intent better to show the two cases - isFirst or not - instead of assuming !isFirst and right after overrdiing both available and offset again if isFirst.  It is probably more performant too.  But it's totally optional whether to do it this way or not.  The main change is the textOverflowed flag - which I admit is a bit hackish.

This whole function is rather complex - I assume with some deep thought it might be possible to express it more clearly.

Also, I made a lot of comments for my own understanding of this function, and I thought perhaps it'd be helpful to try to add these as documentation for others to more easily be able to trace through the flow of this function.  Happy to do so, but as I'm not the original author, it'd be good to have someone that's a bit more familiar do it, or oversee/fix my comments.

Reply:
@majelbstoat I just submitted another commit that adds documentation and renames some variables to make the algorithm more clear, while keeping the logic untouched.

I think it'll help more people understand the algorithm, and therefore empower them to troubleshoot and refactor it.

It's such a fundamental algorithm used by many functions inside of Slate, and it's really rather complicated, so I think it's worth making it more approachable in this way.

Reply:
@majelbstoat It irked me too much, so I refactored the `advance` part of the function to make the logic more clear and remove some smelly code, like the flag.  I'm pretty much done tweaking this function now, so it's safe to code review.  Sorry about the repeated commits.